### ISSUE TITLE
API issue on Windows

### ISSUE THREAD

**Issue by eeholmes** [NOAA-EDAB/demos#10]
@sgaichas @andybeet Adding what we know.

* Started after 2/29/24. More accurately between 03/07/2024 - 03/08/2024. See [below](https://github.com/NOAA-EDAB/demos/issues/10#issuecomment-1995828255)
* Only on NOAA Windows machines, on VPN or off
* Appearing on all NOAA PCs tested. 
    * R 4.2.0, and Rstudio 2022.07.1.544. Not upgraded either in a long time.
    * R 4.3.1 and Rstudio 2023.06.0.421
    * R 4.2.2 and Rstudio 1.4.1103
    * R 4.3.1 and Rstudio 2023.06.2.561
* Happening with [`remotes`](https://github.com/r-lib/remotes) and [`devtools`](https://github.com/r-lib/devtools/) (which uses GitHub API) but not with [`pak`](https://github.com/r-lib/pak) (which doesn't?). Happening with [`jsonlite`](https://github.com/jeroen/jsonlite) which uses GH API too.
* tested on linux using GitHub codespaces (Ubuntu 22.04.3, R 4.3.2) and `devtools::install_github()` worked fine; works fine on all Macs tested (NOAA or non-NOAA; on VPN or off). Works fine on NEFSC Rstudio containers (Linux ubuntu 22.04) (R 4.3.1)

**Comment by andybeet** [comment]
Error messages:

1. using `remotes`

``` r
remotes::install_github("tidyverse/dplyr")
#> Downloading GitHub repo tidyverse/dplyr@HEAD
#> Error in utils::download.file(url, path, method = method, quiet = quiet,  : 
#>   cannot open URL 'https://api.github.com/repos/tidyverse/dplyr/tarball/HEAD'
```

<sup>Created on 2024-03-13 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

2. using `jsonlite`
``` r
jsonlite::fromJSON("https://api.github.com/repos/tidyverse/dplyr/issues")
#> Warning in open.connection(con, "rb"): URL
#> 'https://api.github.com/repos/tidyverse/dplyr/issues': status was 'SSL peer
#> certificate or SSH remote key was not OK'
#> Error in open.connection(con, "rb"): cannot open the connection to 'https://api.github.com/repos/tidyverse/dplyr/issues'
```

<sup>Created on 2024-03-13 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by eeholmes** [comment]
@andybeet Can you run this from a bash terminal on PC showing the problem? Trying to figure out if it is really R related or something about connecting to  the GH API. That SSL error suggests maybe a certificate issue.
```
URL="https://api.github.com/repos/tidyverse/dplyr/issues"
curl $URL
```

**Comment by andybeet** [comment]
@eeholmes Agreed. 
This made me wonder how `remotes` is trying to authenicate me. So I edited the `remotes` package. It is using my PAT.

Is this what you wanted from the bash terminal?
![image](https://github.com/NOAA-EDAB/demos/assets/22455149/8d276417-afba-4c8a-8f10-a6d22a04aca5)

**Comment by eeholmes** [comment]
That's weird. On my Mac, `curl $URL` spits out the json.

**Comment by eeholmes** [comment]
@andybeet What happens if you do this:
```
curl -k $URL
```
the `-k` says not do the security check.

**Comment by andybeet** [comment]
a hah! I get the json file

**Comment by eeholmes** [comment]
Ok, so this suggests that it might be SSL or AntiVirus. Scroll to the bottom.
https://stackoverflow.com/questions/54938026/curl-unknown-error-0x80092012-the-revocation-function-was-unable-to-check-r

**Comment by eeholmes** [comment]
So we know it is not R related! The antivirus update on Windows might have blocked  api.github url. Or it's a SSL issue.  I think next step is to determine if it is SSL or antivirus related.

**Comment by amandalbradford** [comment]
> @sgaichas @andybeet Adding what we know.
> 
> * Started after 2/29/24.
> * Only on NOAA Windows machines, on VPN or off
> * Appearing on all NOAA PCs tested.
>   
>   * R 4.2.0, and Rstudio 2022.07.1.544. Not upgraded either in a long time.
>   * R 4.3.1 and Rstudio 2023.06.0.421
>   * R 4.2.2 and Rstudio 1.4.1103
>   * R 4.3.1 and Rstudio 2023.06.2.561
> * Happening with [`remotes`](https://github.com/r-lib/remotes) (which uses GitHub API) but not with [`pak`](https://github.com/r-lib/pak) (which doesn't?). Happening with [`jsonlite`](https://github.com/jeroen/jsonlite) which uses GH API too.
> * tested on linux using GitHub codespaces (Ubuntu 22.04.3, R 4.3.2) and `devtools::install_github()` worked fine; works fine on all Macs tested (NOAA or non-NOAA; on VPN or off). Works fine on NEFSC Rstudio containers (Linux ubuntu 22.04) (R 4.3.1)

@eeholmes - Clarifying that `devtools::install_github()` produces the same error. I reported the issue on 3/8, and I recall another user saying it was working fine on 3/7/24. I am using  R 4.3.3 and RStudio 2023.12.1.

**Comment by eeholmes** [comment]
@amandalbradford Yes `devtools` would also hit the `api.github` which is the part that has the error. My best guess now it that it is antivirus related (that or SSL certificate). On your PC my guess is 

in a terminal. this gives and error
```
URL="https://api.github.com/repos/tidyverse/dplyr/issues"
curl $URL
```
while this with the `-k` to stop security checking works.
```
curl -k $URL
```

Correct?

**Comment by sgaichas** [comment]
Solution (we think??) from @k-doering-NOAA, thank you Kathryn!

TLDR:
1. Run write('CURL_SSL_BACKEND=openssl', file = "~/.Renviron", append = TRUE) in your R session.
2. Restart your R session
3. Run curl::curl_version()$ssl_version and confirm the return is OpenSSL/1.1.1m (Schannel)   (note the numbers may be different)
4. Try remotes::install_github() (e.g., devtools::install_github('tidyverse/dplyr') ) . It should work and continue to work in future R sessions.

Full details:
This [curl R package vignette](https://cran.r-project.org/web/packages/curl/vignettes/windows.html) explains the 2 different backends for curl for windows users and how to switch between them. My computer defaulted to secure channel (called "Schannel"), which I discovered by calling curl::curl_version()$ssl_version

I then wrote to my .Renviron file to default to using SSL instead: write('CURL_SSL_BACKEND=openssl', file = "~/.Renviron", append = TRUE)

And then restarted my r session. I double-checked the backend being used by calling curl::curl_version()$ssl_version
The return was "OpenSSL/1.1.1m (Schannel)", confirming OpenSSL was being used (the one not in parentheses). If you have a different version of OpenSSL, the numbers may be different for you in the return.

From there, I was able to use devtools::install_github(), and utils::download.file and jsonlite::fromJSON to download files/info from github.

I think the .Renviron file will mean the backend specified will be set as the backend to use the first time curl is used in a session, so running write('CURL_SSL_BACKEND=openssl', file = "~/.Renviron", append = TRUE) one time per computer should fix the issue.

If you wanted to switch back to the old backend, the line in the .Renviron file could be removed.

EDIT: R UG members, Please feel free to keep reporting if this fix works for you or not! I'm curious if there are multiple issues going on and this doesn't work for some.

**Comment by BBeltz1** [comment]
Just noting that this solution did restore my access to GitHub API via RStudio.

**Comment by k-doering-NOAA** [comment]
Thanks for remembering to post the solution on this thread @sgaichas!

**Comment by andybeet** [comment]
All,
Can you see if this issue has resurfaced?
I am experiencing the same issues again?
Can't access the API

However the fix that we tried before still seems to be in effect.

``` r
curl::curl_version()$ssl_version
#> [1] "OpenSSL/1.1.1k (Schannel)"
```

<sup>Created on 2024-07-02 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by sgaichas** [comment]
I can access right now via mac/Rstudio: `jsonlite::fromJSON('https://api.github.com/repos/tidyverse/dplyr/issues')` returns all the tidyverse issues

**Comment by BBeltz1** [comment]
Looks like I'm unaffected as well (on windows)

**Comment by k-doering-NOAA** [comment]
`jsonlite::fromJSON('https://api.github.com/repos/tidyverse/dplyr/issues')` also works for me on Windows from Rstudio

**Comment by GMaynard1** [comment]
This issue is affecting my ability to access the Posit Connect TEST instance (production works fine). I added the CURL_SSL_BACKEND=openssl line into my R environment in order to be able to download .csv files directly from ERDDAP a little while ago and that was working fine. However, I need to remove the CURL_SSL_BACKEND=openssl in order to connect to the TEST instance. If that line is present, I can only connect to production.
